src/features/pdf/data/generatePdfSagas.ts

```diff
@@ -0,0 +1,209 @@
+  } catch (error) {
+    yield put(PdfActions.generateRejected({ error }));
+  }
+}
+
+function* fetchPdfFormatSaga(): SagaIterator {
+  const layoutSets: ILayoutSets = yield select(layoutSetsSelector);
+  const uiConfig: IUiConfig = yield select(uiConfigSelector);
+  const instance: IInstance = yield select(instanceSelector);
+  const applicationMetadata: IApplicationMetadata = yield select(applicationMetadataSelector);
+
+  const dataGuid = getCurrentTaskDataElementId(applicationMetadata, instance, layoutSets);
+  let pdfFormat: IPdfFormat;
+  if (typeof dataGuid === 'string') {
+    try {
+      pdfFormat = yield call(httpGet, getPdfFormatUrl(instance.id, dataGuid));
+    } catch {
+      pdfFormat = {
+        excludedPages: uiConfig.excludePageFromPdf ?? [],
+        excludedComponents: uiConfig.excludeComponentFromPdf ?? [],
+      };
+    }
+  } else {
+    pdfFormat = {
+      excludedPages: uiConfig.excludePageFromPdf ?? [],
+      excludedComponents: uiConfig.excludeComponentFromPdf ?? [],
+    };
+  }
+  yield put(PdfActions.pdfFormatFulfilled({ pdfFormat }));
+}
+
+/**
+ * Watches for changes in formdata and calls fetchPdfFormat and regenerates pdf layout if the method is set to automatic
+ */
+export function* watchPdfPreviewSaga(): SagaIterator {
+  while (true) {
+    yield race([take(FormDataActions.submitFulfilled), take(PdfActions.pdfStateChanged)]);
+    const method: IPdfMethod = yield select(pdfMethodSelector);
+    if (method == 'auto' && pdfPreviewMode()) {
+      yield call(fetchPdfFormatSaga);
+      yield call(generatePdfSaga);
+    }
+  }
+}
+
+/**
+ * Checks if all necessary data is loaded before signaling that pdf-generation is ready
+ */
+export function* watchPdfReadySaga(): SagaIterator {
+  yield all([
+    take(PartyActions.getPartiesFulfilled),
+    take(LanguageActions.fetchLanguageFulfilled),
+    take(TextResourcesActions.fetchFulfilled),
+    take(OrgsActions.fetchFulfilled),
+    take(OptionsActions.loaded),
+    take(DataListsActions.loaded),
+    take(IsLoadingActions.finishDataTaskIsLoading),
+    race([
+      take(FormLayoutActions.noCalculateTriggerFulfilled),
+      take(FormLayoutActions.calculatePageOrderAndMoveToNextPageFulfilled),
+    ]),
+  ]);
+
+  yield put(PdfActions.pdfReady());
+}
+
+export function* watchInitialPdfSaga(): SagaIterator {
+  while (true) {
+    yield race([
+      all([
+        take(QueueActions.startInitialDataTaskQueueFulfilled),
+        take(FormLayoutActions.fetchFulfilled),
+        take(FormLayoutActions.fetchSettingsFulfilled),
+        take(InstanceDataActions.getFulfilled),
+      ]),
+      take(PdfActions.pdfStateChanged),
+    ]);
+    if (shouldGeneratePdf()) {
+      const layouts: ILayouts = yield select(layoutsSelector);
+      const uiConfig: IUiConfig = yield select(uiConfigSelector);
+      const customPdfLayout = uiConfig.pdfLayoutName ? layouts[uiConfig.pdfLayoutName] : undefined;
+      const method = customPdfLayout ? 'custom' : 'auto';
```

Comment:
Not really any other reason than to keep the logic in PDFView simple. It could obviously get the same layout by only knowing the method and selecting the uiConfig itself.

Reply:
I think that would be a better route. Duplicated state is just a good way to ask for future bugs, because everyone just has to "know" that these things are linked together, and have to make sure they never break or change that link. If it's hard to keep the logic simple, prefer to create tools that can help you deduce the state you need from UiConfig instead. It's trivial to for example make a redux selector that reads UiConfig and returns the two things you want here (`method` and `customPdfLayout`).

Reply:
BTW, that philosophy is my end-goal with the (now somwhat infamous) #377 and #345 as well. When you think hard about it, the whole `uiConfig.repeatingGroups` and mapping of repeating groups is mostly just a reflection of the current `formData` state (i.e., "how many rows are there in this repeating group" is just the number of rows defined in the data model). So that's duplicated state, and that leads to problems (see #695, #571, #544 - those are just the things you've fixed). The entire `updateRepeatingGroupsSaga` should go away, IMO, because it's just a necessary evil when we duplicate and dislocate state.

Reply:
Sorry, I'm unresolving the conversation yet again. Noticed that `method` is still duplicated. I also see that there's a state for `readyForPrint` - how is it different from the component of that name, or is that duplicated as well?